sbndaq-artdaq/Generators/Common/CAENV1730Readout_generator.cc

```diff
@@ -1552,6 +1555,18 @@ bool sbndaq::CAENV1730Readout::readWindowDataBlocks() {
 	- (fTTT_ns - (long)fMeanPollTimeNS >  500000000) * 1000000000
 	- fTimeOffsetNanoSec;
     }
+    else if(fUseTimeTagShiftForTimeStamp){
+      fTTT = uint32_t{header->triggerTimeTag}; // 
+      // TTT is 8 ticks/ns, record length is 2 ticks/ns. See CAEN V1730 manuals for details
+      fTTT_ns = (fTTT*8.0) - (((double)fCAEN.recordLength * 2.0) * ((double)fCAEN.postPercent / 100.0)); //in 1 ns
+      
+      // Scheme borrowed from what Antoni developed for CRT.
+      // See https://sbn-docdb.fnal.gov/cgi-bin/private/DisplayMeeting?sessionid=7783
+      fTS = fMeanPollTime - fMeanPollTimeNS + fTTT_ns
+	+ (fTTT_ns - (long)fMeanPollTimeNS < -500000000) * 1000000000
+	- (fTTT_ns - (long)fMeanPollTimeNS >  500000000) * 1000000000
+	- fTimeOffsetNanoSec;
+    }
```

Comment:
From @aaduszki 
Hi, I'm sorry for responding late.
I have questions about the timestamp calculation. In CRT, the FEBs generate hits continuously. The "poll" indicates the time period covering all hits included in a given fragment. The "mean poll time" is supposed to be the time that is somewhere close to all the hits within the fragment. Also, polls are made very frequently (once per 40 or 80 ns).
As far as I understand the readout logic for V1730s is quite different from that. I think we need a separate document (apart from or instead the linked presentation about CRT) explaining how the timestamps from V1730s can be calculated with the same/similar equations.
In the code I see that the "mean poll time" is calculated based on fTimeDiffPollEnd, fTimeDiffPollBegin, which are set at the beginning and end of readWindowDataBlocks(). Is this function called frequently enough to guarantee that these times are very close to the time when V1730s unit actually digitizes some signal? In particular given in ICARUS sometimes the event rate is as low as 0.05 Hz?
A cosmetic suggestion: declare fTimeDiffPollEnd and fTimeDiffPollBegin as static rather than global, to make it apparent they are not modified outside of the function. See for example t_start and t_end here: https://github.com/SBNSoftware/sbndaq-artdaq/blob/develop/sbndaq-artdaq/Generators/Common/BernCRT_GeneratorBase.cc#L295

Reply:
@mvicenzi I think all the assumptions about how frequently the function is called should be tested (or monitored, e.g. in grafana), as in case they are not true, even occasionally, the PMT timestamps would be completely off.